File: src/bin/day10.rs
//! Day 10: Hoof It
//!
//! <https://adventofcode.com/2024/day/10>

use std::{
    collections::{VecDeque, HashSet},
    fmt::Display
};
use aoc_2024::{Solution, get_grid, neighbors_4};

pub struct Day10;

impl Solution for Day10 {
    const NAME: &'static str = "Hoof It";

    fn part_one<T: Display>(&self, inp: T) -> usize {
        let grid = get_grid(inp);
        let rows = 0..grid.len();
        let cols = 0..grid[0].len();

        let mut count = 0;

        for (start_i, row) in grid.iter().enumerate() {
            for (start_j, &cell) in row.iter().enumerate() {
                if cell == b'0' {
                    let mut to_check = VecDeque::new();
                    to_check.push_back((start_i, start_j, 0));

                    let mut nines = HashSet::new();

                    while let Some((i, j, num)) = to_check.pop_front() {
                        if num == 9 {
                            nines.insert((i, j));
                        } else {
                            for (next_i, next_j) in neighbors_4(i, j) {
                                if rows.contains(&next_i)
                                    && cols.contains(&next_j)
                                    && grid[next_i][next_j] as usize - 48 == num + 1
                                {
                                    to_check.push_back((next_i, next_j, num + 1));
                                }
                            }
                        }
                    }
                    count += nines.len();
                }
            }
        }
        count
    }

    fn part_two<T: Display>(&self, inp: T) -> usize {
        let grid = get_grid(inp);
        let rows = 0..grid.len();
        let cols = 0..grid[0].len();

        let mut count = 0;

        for (start_i, row) in grid.iter().enumerate() {
            for (start_j, &cell) in row.iter().enumerate() {
                if cell == b'0' {
                    let mut to_check = VecDeque::new();
                    to_check.push_back((start_i, start_j, 0));

                    while let Some((i, j, num)) = to_check.pop_front() {
                        if num == 9 {
                            count += 1;
                        } else {
                            for (next_i, next_j) in neighbors_4(i, j) {
                                if rows.contains(&next_i)
                                    && cols.contains(&next_j)
                                    && grid[next_i][next_j] as usize - 48 == num + 1
                                {
                                    to_check.push_back((next_i, next_j, num + 1));
                                }
                            }
                        }
                    }
                }
            }
        }
        count
    }

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 430);
        assert_eq!(p2, 928);
    }
}

fn main() {
    aoc_2024::run_day(10, &Day10);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}

File: aoc-py/solutions/day10.py
"""
Day 10: Hoof It

https://adventofcode.com/2024/day/10
"""
__all__ = ('Day10',)

from typing import ClassVar
from collections import deque

from ..solution import Solution
from ..utils import neighbors_4

class Day10(Solution):
    NAME: ClassVar[str] = 'Hoof It'

    def part_one(self, inp: str) -> int:
        grid = inp.splitlines()
        n_rows = len(grid)
        n_cols = len(grid[0])

        count = 0

        for start_i, row in enumerate(grid):
            for start_j, cell in enumerate(row):
                if cell == '0':

                    to_check = deque([(start_i, start_j, 0)])
                    nines = set()

                    while to_check:
                        i, j, num = to_check.popleft()

                        if num == 9:
                            nines.add((i, j))
                        else:
                            for next_i, next_j in neighbors_4(i, j):
                                if (
                                    next_i in range(n_rows)
                                    and next_j in range(n_cols)
                                    and int(grid[next_i][next_j]) == (next_num := num + 1)
                                ):
                                    to_check.append((next_i, next_j, next_num))
                    count += len(nines)
        return count

    def part_two(self, inp: str) -> int:
        grid = inp.splitlines()
        n_rows = len(grid)
        n_cols = len(grid[0])

        count = 0

        for start_i, row in enumerate(grid):
            for start_j, cell in enumerate(row):
                if cell == '0':

                    to_check = deque([(start_i, start_j, 0)])

                    while to_check:
                        i, j, num = to_check.popleft()

                        if num == 9:
                            count += 1
                        else:
                            for next_i, next_j in neighbors_4(i, j):
                                if (
                                    next_i in range(n_rows)
                                    and next_j in range(n_cols)
                                    and int(grid[next_i][next_j]) == (next_num := num + 1)
                                ):
                                    to_check.append((next_i, next_j, next_num))
        return count

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 430
        assert p2 == 928